pkg/migrator/migrator.go

```diff
@@ -39,115 +39,141 @@ func NewMigrator(client client.Client) migrator {
 }
 
-// Migrate updates existing Subscriptions, deletes the CSCs installed during
+// doMigrate updates existing Subscriptions, deletes the CSCs installed during
 // operator installation as well the stale datastore CSCs when the cluster is
-// migrating from Openshift 4.1.x to Openshift 4.2.x
+// migrating from Openshift 4.1.z to Openshift 4.2.z
+func (m *migrator) doMigrate(operatorNamespace string) []string {
+	installedCscs, errorList := m.updateSubscriptions()
+	if errorList != nil {
+		return errorList
+	}
+	errorList = m.deleteInstalledCscs(installedCscs, operatorNamespace)
+	if errorList != nil {
+		return errorList
+	}
+	errorList = m.deleteDatastoreCscs(operatorNamespace)
+	return errorList
+}
+
+// Migrate is a wrapper for doMigrate so that we can retry the migration in case
+// of failures.
 func (m *migrator) Migrate(operatorNamespace string) {
-	installedCscs := m.updateSubscriptions()
-	m.deleteInstalledCscs(installedCscs, operatorNamespace)
-	m.deleteDatastoreCscs(operatorNamespace)
+	errorList := m.doMigrate(operatorNamespace)
+	if errorList != nil {
+		for _, err := range errorList {
+			m.logger.Errorf(err)
+		}
+		m.logger.Infof("[migration] Failed to migrate")
+		m.retry = m.retry - 1
+		if m.retry > 0 {
+			m.logger.Infof("[migration] Retrying migration %s more time(s)", m.retry)
+			m.Migrate(operatorNamespace)
+		}
+	}
 }
 
 // updateSubscriptions updates the existing Subscriptions'
 // spec.source and spec.sourcenamespace fields. Existing
 // Subscriptions referenced installed CatalogSources, which
 // are updated to reference datastore CatalogSources
 // instead.
-func (m *migrator) updateSubscriptions() []types.NamespacedName {
+func (m *migrator) updateSubscriptions() ([]types.NamespacedName, []string) {
 	var installedCscs []types.NamespacedName
+	var errorList []string
 	options := &client.ListOptions{}
 	options.SetLabelSelector(fmt.Sprintf(builders.CscOwnerNameLabel))
 
 	subscriptions := &olm.SubscriptionList{}
 	// Get the list of existing Subscriptions that have the label "csc-owner-name"
 	err := m.client.List(context.TODO(), options, subscriptions)
 	if err != nil {
-		m.logger.Errorf(fmt.Sprintf("Client error: %s", err.Error()))
-		return []types.NamespacedName{}
+		errorList = append(errorList, fmt.Sprintf("[migration] Client error: %s", err.Error()))
+		return []types.NamespacedName{}, errorList
 	}
-	for _, instance := range subscriptions.Items {
-		installedCscs = append(installedCscs, types.NamespacedName{Name: instance.GetLabels()[builders.CscOwnerNameLabel], Namespace: instance.GetLabels()[builders.CscOwnerNamespaceLabel]})
-		// try to infer the datastore CatalogSource from the Subscription
-		datastoreCs, err := findCatalogSource(&instance, m.client)
+	for _, subscription := range subscriptions.Items {
+		installedCscs = append(installedCscs, types.NamespacedName{Name: subscription.GetLabels()[builders.CscOwnerNameLabel], Namespace: subscription.GetLabels()[builders.CscOwnerNamespaceLabel]})
```

Comment:
- I commented about this earlier too. You cannot assume that these labels exist
- Please make this multi-line

Reply:
Ah must have slipped through the cracks thanks for catching that